### ISSUE TITLE
Unable to use Ember.js in a ThymeLeaf application

### ISSUE THREAD

**Issue by manish-in-java** [thymeleaf/thymeleaf#261]
Ember.js uses Handlebars to allow developers to create reusable HTML code snippets.  Sometimes, these snippets have to be decorated with Handlebar syntax to add dynamic logic.  A typical Handlebar snippet looks such as the following:

```
<script type="text/x-handlebars" data-template-name="tasks">
    <ul>
        <li {{bind-attr class="isCompleted:completed"}}>
        </li>
    </ul>
</script>
```

ThymeLeaf fails to parse this snippet, complaining that `li` is not closed correctly on _line 3_.  We therefore have to resort to doing something such as following:

```
<script type="text/x-handlebars" data-template-name="tasks" th:inline="text">
    <ul th:with="open='&lt;li {{bind-attr class=\'isCompleted:completed\'}}&gt;',close='&lt;/li&gt;'">
        [[${open}]]
        [[${close}]]
    </ul>
</script>
```

However, this creates the following problems:
1. Many a times, we need to use `<script th:inline="javascript">` in these scenarios but since only one of `text` or `javascript` is allowed, such mixing cannot be achieved.
2. The whole JavaScript code becomes littered with these constructs and it becomes quite difficult to read and understand the code.

Is there a better way to do this with ThymeLeaf?  Can such integration be achieved in a better way?

**Comment by danielfernandez** [comment]
Hmmm... if you enclose your template (`<script>` body) inside a `<![CDATA[...]]>` section, would it work?

``` xml
    <script type="text/x-handlebars" data-template-name="tasks">
      <![CDATA[
        <ul>
            <li {{bind-attr class="isCompleted:completed"}}>
            </li>
        </ul>
      ]]>
    </script>
```

Of course, this would mean no thymeleaf code inside your ember.js template, but at least, if ember.js understands the `CDATA` block, you'd have a working client-side template...

**Comment by manish-in-java** [comment]
No, that does not work because I need to have ThymeLeaf code inside the templates.

**Comment by danielfernandez** [comment]
That's unfortunate. Right now I cannot think of a way to make it work, because the SAX parser thymeleaf uses underneath has to parse your template in order to process that thymeleaf code, and those ember.js handlebars structures are definitely not valid markup, so you should always get an error :(

As you can see above, I've queued an improvement for attoparser --which will be the parser used in Thymeleaf 3.0-- in order to allow this kind of structures. But as of Thymeleaf 2.1... I cannot see a solution. Sorry.

**Comment by manish-in-java** [comment]
I found another problem with your suggestion.  If the code is written as:

```
<script type="text/x-handlebars" data-template-name="tasks">
  <![CDATA[
    <ul>
        <li {{bind-attr class="isCompleted:completed"}}>
        </li>
    </ul>
  ]]>
</script>
```

browsers interpret `<![CDATA[<ul>` as `<!-- [CDATA[<ul -->` which spoils the template altogether.

However, after much trial and error, I have found that the following workaround works on all browsers:

```
<script type="text/x-handlebars" data-template-name="tasks">
  <![CDATA[
    <!-- 
           guard comment to prevent browsers from considering
           the previous and next lines as part of a comment
    -->
    <ul>
        <li {{bind-attr class="isCompleted:completed"}}>
        </li>
    </ul>
  <!-- ]]> -->
</script>
```

**Comment by BrandonJohnGrenier** [comment]
I've been able to get ThymeLeaf and Ember.js to work together without having to use CDATA sections. I was running into the same problem, then I remembered reading about ThymeLeaf templates:

http://www.thymeleaf.org/doc/html/Using-Thymeleaf.html

ThymeLeaf supports 6 different templates out-of-the-box. If you're using ThymeLeaf in Spring MVC, the "HTML5" ThymeLeaf template is used by default. I configured ThymeLeaf to use the "Legacy HTML5" template instead, and also needed to make sure that nekohtml (1.9.15 or later) was on the classpath. 

After that change, I've been able to use Ember.js (in particular the handlebars expressions) like you'd expect.
